### DISCUSSION TITLE
Character Creation Systems

### DISCUSSION THREAD

**Discussion by owllex** [evennia/evennia#2460] [General Discussion]
Hey everyone,

I was wondering how others have handled the mechanics of character creation, and also looking for some inspiration. I was envisioning a character creation flow that is experienced mostly in-character, and intermixes snippets of lore about the world with decisions for the player to make about how to customize their character (race, background, stats, etc). When that process is finished, then the character would dropped onto the grid.

I imagine this kind of creation system is nothing too unusual, but I couldn't find any good examples or reference points (the chargen contrib isn't quite what I'm looking for).

It seems like what I want could be handled by an EvMenu that echoes text at you in the callbacks between menu options. I was also thinking that it could be done by attaching custom chargen commands to the character and holding them somewhere until they complete the process.

With the EvMenu approach, where is a good place to actually trigger that menu? For example, could it be done in Character.at_post_puppet if we detect you are puppetting a new character? What sort of state is a freshly-made character in at this point? Maybe it should go somewhere else?

With the command approach, I could build a chargen zone that they could interact with. Though this has the downside of potentially causing trouble in a fresh dev instance which don't yet have those rooms built, (I suppose any system that depends on a specific room existing aside from Limbo might have this problem). Setting their location to None might sidestep that problem, but do puppetted characters have an issue with not having a location? Is this approach just messier than it needs to be?

There are probably a bunch of other ideas out there too, I'd love to hear what others have done!

Thanks!

**Comment by davewiththenicehat** [comment]
There is a chargen contrib. I have not read it but I bet it has some good ideas in it.
https://github.com/evennia/evennia/blob/master/evennia/contrib/chargen.py

**Comment by samjaninf** [comment]
I've seen one really interesting creation process that I liked. It was a Smaug MUD. Essentially you would lo in as a new character, then a god or caretaker of souls or some sort of NPC would guide you through character creation. You would be a spirit or something like that on first entry and then the NPC would ask something like, "Do you remember what Race you were?" and it would drop you into the menu to pick a Race. Once you did that he would say something like "Interesting, that's going to make it harder for you. What was your occupation?" or like "What did you look like?" then drop them into the menu to do the thing. Etc etc until they finished. Was too long, but good flavor text and nice bit of In character interactivity. 

your description would also change every time you did something so, you started out as a spirit, then became an Orc looking spirit or osme such, then Red hair, etc etc. Was really well thought out.

**Comment by Griatch** [comment]
> With the EvMenu approach, where is a good place to actually trigger that menu? For example, could it be done in Character.at_post_puppet if we detect you are puppetting a new character? What sort of state is a freshly-made character in at this point? Maybe it should go somewhere else?

There is no need to auto-create a Character with the same name as your Account (it works too though). If you run `MULTISESSION_MODE > 1`, a Character will not be created and you'll log in to an OOCLook that is a little simple menu where you can create/puppet your available characters. That's a great place to start your EvMenu from. If you don't want a higher `MULTISESSION_MODE` you can customize the `UnloggedinCreate` command to not create a Character and then tweak your `Account.at_login` hook to drop you into your Character creation as an Account - that way you can create the Character at the end of your chargen (and aborting chargen will just leave you in OOC mode without any half-finished character hanging around). 

> With the command approach, I could build a chargen zone that they could interact with. Though this has the downside of potentially causing trouble in a fresh dev instance which don't yet have those rooms built, (I suppose any system that depends on a specific room existing aside from Limbo might have this problem). Setting their location to None might sidestep that problem, but do puppetted characters have an issue with not having a location? Is this approach just messier than it needs to be?

Puppeted objects without a location will give an error - this is not a normal state. I would suggest that you customize your login process to generate the chargen zone on the fly and drop the character in it (since supposedly it's instanced and not shared between players).
